### PR TITLE
Enable metric collection for Async Instruments - Delta and Cumulative 

### LINKED ISSUE #1236: [Metric SDK] Cumulative metric collection for Asynchronous Aggregation storage
Refer to [data model ](https://github.com/open-telemetry/opentelemetry-specification/blob/10cac7e35ae4fcf92ed020befe85b3701f4d19e1/specification/metrics/datamodel.md#metric-points) for delta and cumulative metric data points.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1236 #1234 

## Changes

 AsyncMetricStorage invokes the instrumented observation callback and waits for it to be done. After the observation completes, it resets the active storage stash and builds metric data for the metric instrument.
Also created TemporalMetricStorage inspired from [otel-java](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/TemporalMetricStorage.java) to store the history of metrics. This TemporalMetricStorage would be later re-used for synchronous storage too.

@ThomsonTan @esigo - Updated the description, so should be good to review now. Thanks

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                    opentelemetry::common::SystemTimestamp sdk_start_ts,
+                    opentelemetry::common::SystemTimestamp collection_ts,
+                    std::shared_ptr<AttributesHashMap> delta_metrics,
+                    nostd::function_ref<bool(MetricData)> callback) noexcept;
+
+private:
+  InstrumentDescriptor instrument_descriptor_;
+
+  // unreported metrics stash for all the collectors
+  std::unordered_map<CollectorHandle *, std::list<std::shared_ptr<AttributesHashMap>>>
```
Comment (esigo) [inline sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h:39]: Why do we need `std::list` here?

File: sdk/src/metrics/state/temporal_metric_storage.cc [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      auto agg = merged_metrics->Get(attributes);
+      if (agg)
+      {
+        merged_metrics->Set(attributes, std::move(agg->Merge(aggregation)));
+      }
+      else
+      {
+        merged_metrics->Set(
+            attributes,
+            std::move(
```
Comment (esigo) [inline sdk/src/metrics/state/temporal_metric_storage.cc:58]: no need for `std::move` here.

File: sdk/src/metrics/state/temporal_metric_storage.cc [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      [&metric_data](const MetricAttributes &attributes, Aggregation &aggregation) {
+        PointDataAttributes point_data_attr;
+        point_data_attr.point_data = aggregation.ToPoint();
+        point_data_attr.attributes = attributes;
+        metric_data.point_data_attr_.push_back(point_data_attr);
+        return true;
+      });
+  return callback(metric_data);
+
+  return true;
```
Comment (esigo) [inline sdk/src/metrics/state/temporal_metric_storage.cc:126]: this can be removed.

**Review by esigo — COMMENTED** [review]
Thanks for the PR.
I had a few minor comments.

File: sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h (hunk 1 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h:39]: This code in TemporalMetricStorage is mostly same as in SyncmetricStorage for maintain unreported metrics and last reported metrics;
 https://github.com/open-telemetry/opentelemetry-cpp/blob/13ac5c30b6f8ae996d05f3c4084c58f952734ac6/sdk/include/opentelemetry/sdk/metrics/state/sync_metric_storage.h#L118

These structures are explained in the Sync Metric storage flow in design doc (https://docs.google.com/document/d/17PiDI91MQy07kkiSJ7lvlnogvMD3FV85g__WEackj1M/edit).

TemporalMetricStroage is supposed to be shared between Async and Sync storage. Once this PR is merged, I will raise another PR to use TemporalMetricStorage for Sync storage.

File: sdk/include/opentelemetry/sdk/metrics/aggregation/default_aggregation.h [hunk 4]
```diff
@@ -90,6 +91,59 @@ class DefaultAggregation
         return DefaultAggregation::CreateAggregation(instrument_descriptor);
     }
   }
+
+  static std::unique_ptr<Aggregation> CloneAggregation(AggregationType aggregation_type,
+                                                       InstrumentDescriptor instrument_descriptor,
+                                                       const Aggregation &to_copy)
+  {
+    const PointType point_data = to_copy.ToPoint();
+    switch (aggregation_type)
```
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/aggregation/default_aggregation.h:100]: thanks fixed.

File: sdk/include/opentelemetry/sdk/metrics/aggregation/drop_aggregation.h [hunk 5]
```diff
@@ -23,6 +23,8 @@ class DropAggregation : public Aggregation
 public:
   DropAggregation() = default;
 
+  DropAggregation(const DropPointData &point_data) {}
```
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/aggregation/drop_aggregation.h]: Thanks fixed.

File: sdk/include/opentelemetry/sdk/metrics/state/async_metric_storage.h [hunk 6]
```diff
@@ -45,30 +47,42 @@ class AsyncMetricStorage : public MetricStorage
...
     // process the read measurements - aggregate and store in hashmap
     for (auto &measurement : ob_res.GetMeasurements())
     {
-      auto agg = DefaultAggregation::CreateAggregation(aggregation_type_, instrument_descriptor_);
-      agg->Aggregate(measurement.second);
-      active_attributes_hashmap_->Set(measurement.first, std::move(agg));
+      auto aggr = DefaultAggregation::CreateAggregation(aggregation_type_, instrument_descriptor_);
+      aggr->Aggregate(measurement.second);
+      auto prev = cumulative_hash_map_->Get(measurement.first);
+      if (prev != nullptr)
```
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/state/async_metric_storage.h]: Fixed.

File: sdk/include/opentelemetry/sdk/metrics/state/async_metric_storage.h [hunk 7]
```diff
@@ -45,30 +47,42 @@ class AsyncMetricStorage : public MetricStorage
...
     }
 
-    // TBD -> read aggregation from hashmap, and perform metric collection
-    MetricData metric_data;
-    if (metric_collection_callback(std::move(metric_data)))
-    {
-      return true;
-    }
-    return false;
+    return temporal_metric_storage_.buildMetrics(collector, collectors, sdk_start_ts, collection_ts,
```
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/state/async_metric_storage.h:74]: Yes, I was planning to revisit protection logic for sync and async separately from this PR. But for now, have protected TemporalMetricStorage::buildmetrics() with lock

File: sdk/include/opentelemetry/sdk/metrics/aggregation/default_aggregation.h [hunk 8]
```diff
@@ -90,6 +91,53 @@ class DefaultAggregation
...
+    {
+      case AggregationType::kDrop:
+        return std::unique_ptr<Aggregation>(new DropAggregation());
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/metrics/aggregation/default_aggregation.h:103]: Probably make assignment of the new Aggregation in the switch and return it at the end of the function? `return` on each case could cause multiple copy of epilog being generated.

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/aggregation/default_aggregation.h:103]: you mean multiple stack unwinding code being generated. I think modern compilers should be smart enough to handle that.

File: sdk/src/metrics/state/temporal_metric_storage.cc (hunk 2 above)
Comment (lalitb) [reply sdk/src/metrics/state/temporal_metric_storage.cc:59]: Good point. done.

File: sdk/src/metrics/state/temporal_metric_storage.cc (hunk 3 above)
Comment (lalitb) [reply sdk/src/metrics/state/temporal_metric_storage.cc]: good catch. removed.

File: sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h [hunk 9]
```diff
@@ -19,21 +19,22 @@ class LongHistogramAggregation : public Aggregation
...
   LongHistogramAggregation(HistogramPointData &&);
+  LongHistogramAggregation(const HistogramPointData &);
 
   void Aggregate(long value, const PointAttributes &attributes = {}) noexcept override;
 
   void Aggregate(double value, const PointAttributes &attributes = {}) noexcept override {}
 
   /* Returns the result of merge of the existing aggregation with delta aggregation with same
    * boundaries */
-  virtual std::unique_ptr<Aggregation> Merge(const Aggregation &delta) const noexcept override;
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h:29]: Why is virtual removed?

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/aggregation/histogram_aggregation.h:29]: It was not required, as we won't need to subclass `HistogramAggregation` and other aggregations further.

File: sdk/src/metrics/state/temporal_metric_storage.cc [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  // Iterate over the unreporter metrics for `collector` and store result in `merged_metrics`
+  std::unique_ptr<AttributesHashMap> merged_metrics(new AttributesHashMap);
+  for (auto &agg_hashmap : unreported_list)
+  {
+    agg_hashmap->GetAllEnteries(
+        [&merged_metrics, this](const MetricAttributes &attributes, Aggregation &aggregation) {
+          auto agg = merged_metrics->Get(attributes);
+          if (agg)
+          {
+            merged_metrics->Set(attributes, std::move(agg->Merge(aggregation)));
```
Comment (esigo) [inline sdk/src/metrics/state/temporal_metric_storage.cc:53]: nit:
```suggestion
            merged_metrics->Set(attributes, agg->Merge(aggregation));
```

File: sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h (hunk 1 above)
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h:39]: Thanks for the clarification.

**Review by esigo — APPROVED** [review]
LGTM
Thanks for the PR :)

**Pull request merged**